osf_pigeon/pigeon.py

```diff
@@ -0,0 +1,243 @@
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+        get_and_write_file_data_to_temp(
+            f'{settings.OSF_FILES_URL}v1/resources/{guid}/providers/osfstorage/?zip=',
+            temp_dir,
+            'archived_files.zip'
+        )
+        get_and_write_json_to_temp(
+            f'{settings.OSF_API_URL}v2/registrations/{guid}/wikis/',
+            temp_dir,
+            'wikis.json'
+        )
+        get_and_write_json_to_temp(
+            f'{settings.OSF_API_URL}v2/registrations/{guid}/logs/',
+            temp_dir,
+            'logs.json'
+        )
+        get_and_write_json_to_temp(
+            f'{settings.OSF_API_URL}v2/guids/{guid}',
+            temp_dir,
+            'registration.json'
+        )
+
+        bag_and_tag(
+            temp_dir,
+            guid,
+            datacite_username=datacite_username,
+            datacite_password=datacite_password,
+            datacite_prefix=datacite_prefix
+        )
+
+        zip_data = create_zip_data(temp_dir)
+
+        assert isinstance(ia_access_key, str),\
+            'Internet Archive access key was not passed to pigeon'
+        assert isinstance(ia_secret_key, str), \
+            'Internet Archive secret key not passed to pigeon'
+
+        session = internetarchive.get_session(
+            config={
+                's3': {
+                    'access': ia_access_key,
+                    'secret': ia_secret_key
+                }
+            }
+        )
+        ia_item = session.get_item(guid)
+        metadata = get_metadata(temp_dir, 'registration.json')
+
+        ia_item.upload(
+            {'bag.zip': zip_data},
+            metadata=metadata,
+            headers={'x-archive-meta01-collection': settings.OSF_COLLECTION_NAME},
+            access_key=ia_access_key,
+            secret_key=ia_secret_key,
+        )
+        modify_metadata_with_retry(ia_item, metadata)
+
+
+def build_doi(guid):
+    return settings.DOI_FORMAT.format(prefix=settings.DATACITE_PREFIX, guid=guid)
+
+
+def get_datacite_metadata(doi, datacite_username, datacite_password, datacite_prefix):
+    assert isinstance(datacite_password, str), 'Datacite password not passed to pigeon'
+    assert isinstance(datacite_username, str), 'Datacite username not passed to pigeon'
+    assert isinstance(datacite_prefix, str), 'Datacite prefix not passed to pigeon'
+    client = DataCiteMDSClient(
+        url=settings.DATACITE_URL,
+        username=datacite_username,
+        password=datacite_password,
+        prefix=datacite_prefix,
+    )
+    return client.metadata_get(doi)
+
+
+@sleep_and_retry
+def get_with_retry(
+        url,
+        retry_on: Tuple[int] = (),
+        sleep_period: int = None,
+        headers: Dict = None) -> requests.Response:
+
+    if not headers:
+        headers = {}
+
+    if not settings.OSF_THROTTLE_ENABLED:
+        assert settings.OSF_BEARER_TOKEN, 'must have OSF_BEARER_TOKEN set to disable throttle'
```

Comment:
Technically this will just make the throttling less likely. You could still be throttled if you're using a token.